apps/glusterfs/volume_entry_allocate.go

```diff
@@ -63,6 +66,209 @@ func (v *VolumeEntry) allocBricksInCluster(db *bolt.DB,
+		brickEntries = append(brickEntries, brick)
+	}
+
+	for _, brickentry := range brickEntries {
+		nodeEntry, err := NewNodeEntryFromId(tx, brickentry.Info.NodeId)
+		if err != nil {
+			return nil, err
+		}
+		if brickname == fmt.Sprintf("%v:%v", nodeEntry.Info.Hostnames.Storage[0], brickentry.Info.Path) {
+			return brickentry, nil
+		}
+	}
+	return nil, ErrNotFound
+}
+
+func (v *VolumeEntry) replaceBrickInVolume(db *bolt.DB, executor executors.Executor,
+	allocator Allocator,
+	oldBrickId string) (e error) {
+
+	logger.Info("Entered replace brick")
+	// Do the work in the database context so that the cluster
+	// data does not change while determining brick location
+	err := db.Update(func(tx *bolt.Tx) error {
+
+		oldBrickEntry, err := NewBrickEntryFromId(tx, oldBrickId)
+		if err != nil {
+			return err
+		}
+
+		oldDeviceEntry, err := NewDeviceEntryFromId(tx, oldBrickEntry.Info.DeviceId)
+		if err != nil {
+			return err
+		}
+		oldBrickNode, err := NewNodeEntryFromId(tx, oldBrickEntry.Info.NodeId)
+		if err != nil {
+			return err
+		}
+
+		// Determine the setlist by getting data from Gluster
+		vinfo, err := executor.VolumeInfo(oldBrickNode.ManageHostName(), v.Info.Name)
+		var slicestartindex int
+		var foundbrickset bool
+		var brick executors.Brick
+		setlist := make([]*BrickEntry, 0)
+		for slicestartindex = 0; slicestartindex <= len(vinfo.Bricks.Bricks)-v.Durability.BricksInSet(); slicestartindex = slicestartindex + v.Durability.BricksInSet() {
+			setlist = make([]*BrickEntry, 0)
+			for _, brick = range vinfo.Bricks.Bricks[slicestartindex : slicestartindex+v.Durability.BricksInSet()] {
+				brickentry, err := v.getEntryfromBrickName(tx, brick.Name)
+				if err != nil {
+					return err
+				}
+				if brickentry.Id() == oldBrickId {
+					foundbrickset = true
+				} else {
+					setlist = append(setlist, brickentry)
+				}
+			}
+			if foundbrickset {
+				break
+			}
+		}
+		if !foundbrickset {
+			return err
+		}
+
+		for _, brickInSet := range setlist {
+			logger.Info("setlist is %v", brickInSet)
+		}
+
+		//Create an Id for new brick
+		var newBrickId string
+		newId := false
+		for !newId {
+			newBrickId = utils.GenUUID()
+			if !utils.SortedStringHas(v.BricksIds(), newBrickId) {
+				newId = true
+			}
+		}
+
+		logger.Info("Ask allocator for the list of devices")
+		// Check the ring for devices to place the brick
+		deviceCh, done, errc := allocator.GetNodes(v.Info.Cluster, newBrickId)
+		defer func() {
+			close(done)
+		}()
+
+		for deviceId := range deviceCh {
+
+			// Get device entry
+			newDeviceEntry, err := NewDeviceEntryFromId(tx, deviceId)
+			if err != nil {
+				return err
+			}
+
+			// Skip same device
+			if oldDeviceEntry.Info.Id == newDeviceEntry.Info.Id {
+				continue
+			}
+
+			// Do not allow a device from the same node to be
+			// in the set
+			deviceOk := true
+			for _, brickInSet := range setlist {
+				if brickInSet.Info.NodeId == newDeviceEntry.NodeId {
+					deviceOk = false
+				}
+			}
+
```

Comment:
You have brought up a good point, but the allocator does not do that.  Which actually, is a bigger issue because this is also a problem with the normal allocation of bricks, where they could go into the same zone.

The allocator creates an array of the following form for a topology of three zones: [z1d0n0 z2d0n0, z3d0n0, z1d1n0, z2d1n0, z3d1n0, z1d0n1, z3d1n0, z3d1n0, and so on..].  So if z1d0n0 is the disk that needs to be replaced, it may choose z3d0n0.  But in the set for z1d0n0, it could have [z1d0n0, z2d1n0, z3d2n1], and if it chose z3d0n0 for replacing z1d0n0, now it has two drives in the same zone.  Instead it should make sure to go through the drives looking _first_ for drives _not_ in a zone already in the set.  Then if it does not find one because they have no space, then we may have to pick one in a zone which is already used as a fallback.

This sounds like it could be an issue also in the normal allocation, so solving it here we may use it as a fix for the normal allocation.

I'm not opposed to pushing this to its own fix after this if you want were we can concentrate the fix for both allocations in that one fix and provide the appropriate tests.  Up to you.